### ISSUE TITLE
 Reducing CPU consumption when running multiple instances of YOLOv8 using multithreading and multiprocessing

### ISSUE THREAD

**Issue by Rasantis** [ultralytics/ultralytics#13604]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have a project that processes 4 videos simultaneously. It works well, but I need to scale these operations. The GPU handles it smoothly, but the CPU usage is around 90%, which makes it impossible to scale this application for more cameras. I am using a Windows VM on AWS with an NVIDIA T4 GPU with 16GB of memory. The machine I am using is a g4dn.2xlarge. I am using Python 3.11.4 and Ultralytics==8.1.4. Here is my code. If you could take a look and provide feedback on what could be implemented and improved to reduce CPU usage, that would be great!

### Additional

my code fence.py:
import cv2
from ultralytics import YOLO
import winsound
import threading
import numpy as np
import json
import torch
import requests
import os
import datetime
import torch.multiprocessing as mp

# Telegram Bot API Token
telegram_token = "a"

def send_telegram_message(message, chat_id, image_path=None):
    url = f"https://api.telegram.org/bot{telegram_token}/sendMessage"
    params = {"chat_id": chat_id, "text": message}
    if image_path:
        url = f"https://api.telegram.org/bot{telegram_token}/sendPhoto"
        files = {'photo': open(image_path, 'rb')}
        params['caption'] = message
        response = requests.post(url, params=params, files=files)
    else:
        response = requests.post(url, params=params)

    if response.status_code != 200:
        print("Failed to send Telegram message:", response.content)

if torch.cuda.is_available():
    torch.cuda.set_device(0)

class VirtualFence:
    def __init__(self, video_path, model_path, cameras_json_path, camera_name, user_chat_id=None):
        self.video = cv2.VideoCapture(video_path)
        self.modelo = YOLO(model_path).to('cuda' if torch.cuda.is_available() else 'cpu')
        self.display_width = 1100
        self.user_chat_id = user_chat_id  # Novo atributo
        
        with open(cameras_json_path, 'r') as file:
            cameras_data = json.load(file)
        
        camera_data = cameras_data["cameras"].get(camera_name)
        if not camera_data:
            raise ValueError(f"Câmera {camera_name} não encontrada no arquivo {cameras_json_path}")
        
        self.areas = [np.array(v, dtype=np.int32) for v in camera_data["zones"].values() if len(v) >= 3]

        self.original_width = int(self.video.get(cv2.CAP_PROP_FRAME_WIDTH))
        self.original_height = int(self.video.get(cv2.CAP_PROP_FRAME_HEIGHT))
        
        # Adicione a verificação aqui
        if self.original_width == 0 or self.original_height == 0:
            raise ValueError("O vídeo não forneceu dimensões válidas. Verifique se o caminho do vídeo está correto e se o vídeo é acessível.")

        self.alarmeCtl = False

    def alarme(self, image_path):
        for _ in range(7):
            winsound.Beep(2500, 500)
            self.alarmeCtl = False
        if self.user_chat_id:
            send_telegram_message("Invasor detectado", self.user_chat_id, image_path)

    def run(self):
        print(f"Original resolution: ({self.original_width}, {self.original_height})")
        self.display_height = self.display_width * self.original_height // self.original_width
        print(f"Display resolution: ({self.display_width}, {self.display_height})")

        transformation_matrix = cv2.getAffineTransform(np.float32([[0, 0], [self.original_width, 0], [0, self.original_height]]), 
                                                        np.float32([[0, 0], [self.display_width, 0], [0, self.display_height]]))

        timestamp = datetime.datetime.now().strftime("%Y%m%d%H%M%S")
        output_video_path = f"cerca_virtual/interno/output_{timestamp}.mp4"
        
        fourcc = cv2.VideoWriter_fourcc(*'mp4v')
        out = cv2.VideoWriter(output_video_path, fourcc, 30.0, (self.display_width, self.display_height))

        while True:
            check, img = self.video.read()
            if not check:
                break

            img = cv2.resize(img, (self.display_width, self.display_height), interpolation=cv2.INTER_AREA)
            img2 = img.copy()

            resized_areas = [cv2.transform(np.expand_dims(area, axis=1), transformation_matrix).squeeze().astype(np.int32) for area in self.areas]

            for area in resized_areas:
                if len(area) > 0:
                    cv2.fillPoly(img2, [area], color=(0, 255, 0))

            resultado = self.modelo(img, device='cuda' if torch.cuda.is_available() else 'cpu')

            for objetos in resultado:
                obj = objetos.boxes
                for dados in obj:
                    x, y, w, h = map(int, dados.xyxy[0])
                    cls = int(dados.cls[0])
                    cx, cy = (x + w) // 2, (y + h) // 2
                    if cls == 0:
                        cv2.rectangle(img, (x, y), (w, h), (255, 0, 0), 5)
                        for area in resized_areas:
                            if len(area) > 0 and cv2.pointPolygonTest(area, (cx, cy), False) >= 0:
                                cv2.fillPoly(img2, [area], color=(0, 0, 255))
                                cv2.rectangle(img, (100, 30), (470, 80), (0, 0, 255), -1)
                                cv2.putText(img, 'INVASOR DETECTADO', (105, 65), cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 255), 3)

            imgFinal = cv2.addWeighted(img2, 0.5, img, 0.5, 0)

            if not self.alarmeCtl:
                image_path = "cerca_virtual/interno/image.jpg"  # Substitua pelo local onde você quer salvar a imagem
                cv2.imwrite(image_path, imgFinal)
                self.alarmeCtl = True
                threading.Thread(target=self.alarme, args=(image_path,)).start()

            out.write(imgFinal)
            cv2.imshow('img', imgFinal)
            if cv2.waitKey(1) & 0xFF == ord('q'):
                break

        self.video.release()
        out.release()
        cv2.destroyAllWindows()

def start_virtual_fence(camera_name, cameras_json_path, model_path, user_chat_id):
    with open(cameras_json_path, 'r') as file:
        cameras_data = json.load(file)
        
    camera_data = cameras_data["cameras"].get(camera_name)
    if not camera_data:
        raise ValueError(f"Câmera {camera_name} não encontrada no arquivo {cameras_json_path}")

    video_path = camera_data.get("video_path")
    if not video_path:
        raise ValueError(f"Caminho do vídeo para a câmera {camera_name} não especificado no arquivo {cameras_json_path}")

    virtual_fence = VirtualFence(video_path, model_path, cameras_json_path, camera_name, user_chat_id)
    virtual_fence.run()

def run_virtual_fences():
    cameras_json_path = 'cerca_virtual/interno/cameras.json'
    model_path = 'modelo/yolov8n.pt'
    user_chat_id = 'user_chat_id'  # Adapte conforme necessário

    with open(cameras_json_path, 'r') as file:
        cameras_data = json.load(file)

    camera_names = list(cameras_data["cameras"].keys())

    with mp.Pool(processes=len(camera_names)) as pool:
        pool.starmap(start_virtual_fence, [(camera_name, cameras_json_path, model_path, user_chat_id) for camera_name in camera_names])

# Se você quiser rodar diretamente este script, descomente a linha abaixo.
if __name__ == "__main__":
    run_virtual_fences()

# IM SENDING MY JSON FILE TOO 
[cameras.json](https://github.com/user-attachments/files/15844290/cameras.json)

**Comment by glenn-jocher** [comment]
Hi there,

Thank you for sharing your detailed code and setup. To help reduce CPU consumption when running multiple instances of YOLOv8, here are a few suggestions:

1. **Ensure Latest Versions**: First, please make sure you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using:
    ```bash
    pip install --upgrade torch ultralytics
    ```

2. **Optimize Data Loading**: Use efficient data loading techniques. For example, you can use `cv2.VideoCapture` in a separate thread to preload frames.

3. **Reduce CPU Usage with Multiprocessing**: Instead of using threading, which can be CPU-intensive due to Python's Global Interpreter Lock (GIL), consider using multiprocessing for parallel processing. This can help distribute the load across multiple CPU cores.

4. **Batch Processing**: If possible, process frames in batches rather than one by one. This can significantly reduce the overhead.

5. **Model Inference on GPU**: Ensure that all model inferences are performed on the GPU. This can be done by setting the model to use the GPU explicitly:
    ```python
    self.modelo = YOLO(model_path).to('cuda')
    ```

6. **Efficient Video Processing**: Use more efficient video codecs and formats that are less CPU-intensive.

Here's a revised version of your `VirtualFence` class with multiprocessing and some optimizations:

```python
import cv2
from ultralytics import YOLO
import winsound
import multiprocessing as mp
import numpy as np
import json
import torch
import requests
import os
import datetime

# Telegram Bot API Token
telegram_token = "a"

def send_telegram_message(message, chat_id, image_path=None):
    url = f"https://api.telegram.org/bot{telegram_token}/sendMessage"
    params = {"chat_id": chat_id, "text": message}
    if image_path:
        url = f"https://api.telegram.org/bot{telegram_token}/sendPhoto"
        files = {'photo': open(image_path, 'rb')}
        params['caption'] = message
        response = requests.post(url, params=params, files=files)
    else:
        response = requests.post(url, params=params)

    if response.status_code != 200:
        print("Failed to send Telegram message:", response.content)

if torch.cuda.is_available():
    torch.cuda.set_device(0)

class VirtualFence:
    def __init__(self, video_path, model_path, cameras_json_path, camera_name, user_chat_id=None):
        self.video = cv2.VideoCapture(video_path)
        self.modelo = YOLO(model_path).to('cuda' if torch.cuda.is_available() else 'cpu')
        self.display_width = 1100
        self.user_chat_id = user_chat_id
        
        with open(cameras_json_path, 'r') as file:
            cameras_data = json.load(file)
        
        camera_data = cameras_data["cameras"].get(camera_name)
        if not camera_data:
            raise ValueError(f"Câmera {camera_name} não encontrada no arquivo {cameras_json_path}")
        
        self.areas = [np.array(v, dtype=np.int32) for v in camera_data["zones"].values() if len(v) >= 3]

        self.original_width = int(self.video.get(cv2.CAP_PROP_FRAME_WIDTH))
        self.original_height = int(self.video.get(cv2.CAP_PROP_FRAME_HEIGHT))
        
        if self.original_width == 0 or self.original_height == 0:
            raise ValueError("O vídeo não forneceu dimensões válidas. Verifique se o caminho do vídeo está correto e se o vídeo é acessível.")

        self.alarmeCtl = False

    def alarme(self, image_path):
        for _ in range(7):
            winsound.Beep(2500, 500)
            self.alarmeCtl = False
        if self.user_chat_id:
            send_telegram_message("Invasor detectado", self.user_chat_id, image_path)

    def run(self):
        print(f"Original resolution: ({self.original_width}, {self.original_height})")
        self.display_height = self.display_width * self.original_height // self.original_width
        print(f"Display resolution: ({self.display_width}, {self.display_height})")

        transformation_matrix = cv2.getAffineTransform(np.float32([[0, 0], [self.original_width, 0], [0, self.original_height]]), 
                                                        np.float32([[0, 0], [self.display_width, 0], [0, self.display_height]]))

        timestamp = datetime.datetime.now().strftime("%Y%m%d%H%M%S")
        output_video_path = f"cerca_virtual/interno/output_{timestamp}.mp4"
        
        fourcc = cv2.VideoWriter_fourcc(*'mp4v')
        out = cv2.VideoWriter(output_video_path, fourcc, 30.0, (self.display_width, self.display_height))

        while True:
            check, img = self.video.read()
            if not check:
                break

            img = cv2.resize(img, (self.display_width, self.display_height), interpolation=cv2.INTER_AREA)
            img2 = img.copy()

            resized_areas = [cv2.transform(np.expand_dims(area, axis=1), transformation_matrix).squeeze().astype(np.int32) for area in self.areas]

            for area in resized_areas:
                if len(area) > 0:
                    cv2.fillPoly(img2, [area], color=(0, 255, 0))

            resultado = self.modelo(img, device='cuda' if torch.cuda.is_available() else 'cpu')

            for objetos in resultado:
                obj = objetos.boxes
                for dados in obj:
                    x, y, w, h = map(int, dados.xyxy[0])
                    cls = int(dados.cls[0])
                    cx, cy = (x + w) // 2, (y + h) // 2
                    if cls == 0:
                        cv2.rectangle(img, (x, y), (w, h), (255, 0, 0), 5)
                        for area in resized_areas:
                            if len(area) > 0 and cv2.pointPolygonTest(area, (cx, cy), False) >= 0:
                                cv2.fillPoly(img2, [area], color=(0, 0, 255))
                                cv2.rectangle(img, (100, 30), (470, 80), (0, 0, 255), -1)
                                cv2.putText(img, 'INVASOR DETECTADO', (105, 65), cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 255), 3)

            imgFinal = cv2.addWeighted(img2, 0.5, img, 0.5, 0)

            if not self.alarmeCtl:
                image_path = "cerca_virtual/interno/image.jpg"
                cv2.imwrite(image_path, imgFinal)
                self.alarmeCtl = True
                mp.Process(target=self.alarme, args=(image_path,)).start()

            out.write(imgFinal)
            cv2.imshow('img', imgFinal)
            if cv2.waitKey(1) & 0xFF == ord('q'):
                break

        self.video.release()
        out.release()
        cv2.destroyAllWindows()

def start_virtual_fence(camera_name, cameras_json_path, model_path, user_chat_id):
    with open(cameras_json_path, 'r') as file:
        cameras_data = json.load(file)
        
    camera_data = cameras_data["cameras"].get(camera_name)
    if not camera_data:
        raise ValueError(f"Câmera {camera_name} não encontrada no arquivo {cameras_json_path}")

    video_path = camera_data.get("video_path")
    if not video_path:
        raise ValueError(f"Caminho do vídeo para a câmera {camera_name} não especificado no arquivo {cameras_json_path}")

    virtual_fence = VirtualFence(video_path, model_path, cameras_json_path, camera_name, user_chat_id)
    virtual_fence.run()

def run_virtual_fences():
    cameras_json_path = 'cerca_virtual/interno/cameras.json'
    model_path = 'modelo/yolov8n.pt'
    user_chat_id = 'user_chat_id'

    with open(cameras_json_path, 'r') as file:
        cameras_data = json.load(file)

    camera_names = list(cameras_data["cameras"].keys())

    with mp.Pool(processes=len(camera_names)) as pool:
        pool.starmap(start_virtual_fence, [(camera_name, cameras_json_path, model_path, user_chat_id) for camera_name in camera_names])

if __name__ == "__main__":
    run_virtual_fences()
```

This script uses multiprocessing to handle multiple video streams, which should help reduce CPU usage. Additionally, ensure that all model inferences are performed on the GPU to offload the CPU.

For more details on tracking and configuration, please refer to our [tracking documentation](https://docs.ultralytics.com/modes/track/).

Feel free to reach out if you have any further questions! 😊

**Comment by Rasantis** [comment]
huum, nice, and about processing the frames in batch, i saw in the docs that the training method its possible to train with a batch size, but to predict i believe not, is there an a example on how could i try processing in battches the frames in my code....?

**Comment by glenn-jocher** [comment]
@Rasantis hello,

Thank you for your question! Processing frames in batches during prediction can indeed help optimize performance. While the training method explicitly supports batch sizes, you can also achieve batch processing during prediction by leveraging the `predict` method in a loop to handle multiple frames at once.

Here's an example of how you can modify your code to process frames in batches:

```python
import cv2
from ultralytics import YOLO
import torch

# Load the YOLOv8 model
model = YOLO("yolov8n.pt").to('cuda' if torch.cuda.is_available() else 'cpu')

def process_batch(frames):
    # Run inference on a batch of frames
    results = model(frames)
    return results

def main(video_path):
    cap = cv2.VideoCapture(video_path)
    batch_size = 4  # Adjust batch size as needed
    frames = []

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        frames.append(frame)
        if len(frames) == batch_size:
            results = process_batch(frames)
            for result in results:
                annotated_frame = result.plot()
                cv2.imshow("YOLOv8 Inference", annotated_frame)
                if cv2.waitKey(1) & 0xFF == ord('q'):
                    break
            frames = []

    cap.release()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main("path/to/your/video.mp4")
```

This script processes video frames in batches of 4. Adjust the `batch_size` as needed based on your hardware capabilities.

Please ensure you are using the latest versions of `torch` and `ultralytics` to benefit from the latest features and optimizations. You can upgrade them using:
```bash
pip install --upgrade torch ultralytics
```

If you encounter any issues, please provide a minimum reproducible code example so we can assist you better. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any further questions! 😊

**Comment by Y-T-G** [comment]
Don't do it this way. Use Triton Inference Server.

**Comment by glenn-jocher** [comment]
Hello @Y-T-G,

Thank you for your suggestion! Using Triton Inference Server is indeed a powerful way to manage and scale model inference, especially for high-performance applications. Triton can handle multiple models, batch requests, and optimize GPU utilization, which can significantly reduce CPU load and improve overall performance.

For those interested in exploring Triton Inference Server, you can find more information and setup instructions in the [NVIDIA Triton Inference Server documentation](https://github.com/triton-inference-server/server).

If you encounter any issues or have specific questions about integrating Triton with YOLOv8, please provide a minimum reproducible code example. This will help us better understand your setup and provide more targeted assistance. You can find guidelines for creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, ensure you are using the latest versions of `torch` and `ultralytics` to benefit from the latest features and optimizations. You can upgrade them using:
```bash
pip install --upgrade torch ultralytics
```

Feel free to reach out if you have any further questions or need additional assistance! 😊
